### PR TITLE
Dev/rebuild sandbox

### PR DESCRIPTION
**Pull request by BeachWang**
1. Separation of sanbox and data processing parameters;
2. pipeline is automatically executed according to the configuration file list sequence.
3. The parsing of job parameters is adjusted to the execution before job execution, and the refine of recipes is supported for use in the current round;
4. Since Ksigma used to store all init parameters, it cannot be used as the DJ parameter file. Instead, only the parameters in the configuration file with a low value are stored.

Minor:
1. The subconfiguration file parsing function prepare_side_configs in data_juicer/config/config.py, this part of the function may be repeated with gece's parameter reconfiguration;
2. The HPO part of the sandbox is not modified. It needs to be checked for correctness.

### CODE REVIEW

**Comment by drcege** [comment]
TODO: 
- [ ] More discussions on major refactoring
- [ ] Align with the concepts in the paper
- [ ] Correct minor errors

File: data_juicer/core/sandbox/hooks.py [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+        self.specify_dj_and_extra_configs()
+        model_evaluator = model_evaluator_factory(self.other_cfg)
+        # basic routine to evaluate the given model,
+        # users can customize this freely
+        logger.info('Begin to evaluate the model with given evaluator config')
+        model_evaluator.run(kwargs)
+        return kwargs
+
+
+HOOK_DICT = {
```
Comment (drcege) [inline data_juicer/core/sandbox/hooks.py:252]: Usually `MAPPING`

File: data_juicer/core/sandbox/hooks.py [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+    'RefineRecipeViaKSigmaHook': RefineRecipeViaKSigmaHook,
+    'RefineRecipeViaModelFeedbackHook': RefineRecipeViaModelFeedbackHook,
+    'ProcessDataHook': ProcessDataHook,
+    'TrainModelHook': TrainModelHook,
+    'EvaluateDataHook': EvaluateDataHook,
+    'EvaluateModelHook': EvaluateModelHook,
+}
+
+
+def regist_hook(job_cfg, watcher):
```
Comment (drcege) [inline data_juicer/core/sandbox/hooks.py:264]: Should be `register`

File: configs/demo/sandbox/sandbox.yaml [hunk 3]
```diff
@@ -1,27 +1,68 @@
...
-dataset_path: './demos/data/demo-dataset.jsonl'  # path to your dataset directory or file
-np: 4  # number of subprocess to process your dataset
+experiment_name: 'demo-sandbox-run0'              # for wandb tracer name
+hpo_config: null                                  # path to a configuration file when using auto-HPO tool.
 
-export_path: './outputs/demo-sandbox/demo-sandbox.jsonl'
+# configs for each job, the jobs will be executed according to the order in the list
+probe_job_configs:
+  - hook: 'ProbeViaAnalyzerHook'
+    meta_name: 'analysis_ori_data'
```
Comment (drcege) [inline configs/demo/sandbox/sandbox.yaml:11]: How about `label` or something easy to understand?

File: demos/data_mixture/data/the-pile-nih-refined.jsonl [hunk 4]
```diff
@@ -5,6 +5,6 @@
...
+{"text":"This research seeks to understand how and why households enter and exit material hardship, the extent to which federal program participation acts as a buffer, and how material hardship, not poverty, affects children and adults over the long-term. Specifically, the research will: (1) Identify how individual-level characteristics (e.g., family type, veteran status) and decisions impact the experience of material hardship. (2) Document how structural determinants at the state level (e.g., use of technology in application procedures) influence individual well-being. (3) Evaluate the consequences of material hardship over the life course (e.g., fertility, college attendance). Econometric techniques will be applied to data from the 2008 panel of the Survey of Income and Program Participation (SIPP) and longitudinal data from the Panel Study of Income Dynamics (PSID). Models of material hardship will be developed based on four domains of need: home, medical, bill-paying, and food hardship. Analyzes will include descriptive analyzes and individual fixed effects models."}
```
Comment (drcege) [inline demos/data_mixture/data/the-pile-nih-refined.jsonl:8]: what changed?

File: demos/process_sci_data/data/arxiv.jsonl [hunk 5]
```diff
@@ -6,6 +6,6 @@
...
...8ru} for $^{208}$Pb, but close to results in \\cite{Yasui:2012rw}. \n\nThe experimental detection of $D$ and $\\bar D$-meson bound states is, though, a difficult task \\cite{Yamagata-Sekihara:2015ebw}. The formation of D-mesic nuclei with antiprotons beams at PANDA (FAIR) might be possible if ($\\bar p$, D+N) or ($\\bar p$, D+2N) reactions with small or even zero momentum transfer are produced, although the formation cross sections could be suppressed because of the complexity of the reaction mechanisms. Other competing mechanisms could involve the emission of pions by intermediate $D^*$ or $\\bar D^*$ with subsequent trapping of the pseudoscalar charmed mesons by the final nucleus \\cite{Yamagata-Sekihara:2015ebw}.\n\n\\section*{Acknowledgements}\nL.T. acknowledges support from the Heisenberg Programme of the Deutsche Forschungsgemeinschaft under the Project Nr. 383452331,  the Ram\\'on y Cajal research programme, FPA2013-43425-P and FPA2016-81114-P Grants from MINECO, and THOR COST Action CA15213.\n\n\n\\bibliographystyle{JHEP}\n","meta":{"arxiv_id":"1711.10711","language":"en","source":"arxiv","timestamp":1512032756000,"url":"https:\/\/arxiv.org\/abs\/1711.10711","yymm":"1711"}}
```
Comment (drcege) [inline demos/process_sci_data/data/arxiv.jsonl:9]: what changed?

Comment (BeachWang) [reply demos/process_sci_data/data/arxiv.jsonl:9]: analyse

**Review by yxdyc — APPROVED** [review]
LGTM

**Pull request merged**